internal/docker/volumes.go

```diff
@@ -0,0 +1,166 @@
+	"bufio"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/docker/docker/api/types/mount"
+	"github.com/docker/docker/api/types/volume"
+	client "github.com/docker/docker/client"
+)
+
+// A volume is a mounted directory on the homeserver container
+type Volume interface {
+	// Return the mount point
+	Mount() mount.Mount
+	// Prepare the mount point. `contextStr` is unique to this blueprint+homeserver for homeserver
+	// specific mounts e.g appservices.
+	Prepare(ctx context.Context, docker *client.Client, contextStr string) error
+}
+
+type VolumeCA struct {
+	source string
+	typ    mount.Type
+}
+
+// Prepare the Certificate Authority volume. This is independent of the homeserver calling Prepare
+// hence the contextual string is unused.
+func (v *VolumeCA) Prepare(ctx context.Context, docker *client.Client, x string) error {
+	// TODO: wrap in a lockfile
+	if os.Getenv("CI") == "true" {
+		// When in CI, Complement itself is a container with the CA volume mounted at /ca.
+		// We need to mount this volume to all homeserver containers to synchronize the CA cert.
+		// This is needed to establish trust among all containers.
+
+		containerID := getContainerID()
+		if containerID == "" {
+			return fmt.Errorf("failed to get container ID")
+		}
+		container, err := docker.ContainerInspect(ctx, containerID)
+		if err != nil {
+			return err
+		}
+		// Get the volume that matches the destination in our complement container
+		for i := range container.Mounts {
+			if container.Mounts[i].Destination == "/ca" {
+				v.source = container.Mounts[i].Name
+				v.typ = container.Mounts[i].Type
+				break
+			}
+		}
+		if v.source == "" {
+			// We did not find a volume. This container might be created without a volume,
+			// or CI=true is passed but we are not running in a container.
+			return fmt.Errorf("CI=true but there is no /ca mounted to Complement's container")
+		}
+	} else {
+		// When not in CI, our CA cert is placed in the current working dir.
+		// We bind mount this directory to all homeserver containers.
+		cwd, err := os.Getwd()
+		if err != nil {
+			return err
+		}
+		caCertificateDirHost := path.Join(cwd, "ca")
+		if _, err := os.Stat(caCertificateDirHost); os.IsNotExist(err) {
+			err = os.Mkdir(caCertificateDirHost, 0770)
+			if err != nil {
+				return err
+			}
+		}
+		v.source = path.Join(cwd, "ca")
+	}
+	return nil
+}
+
+func (v *VolumeCA) Mount() mount.Mount {
+	return mount.Mount{
+		Type:   v.typ,
+		Source: v.source,
+		Target: "/ca",
+	}
+}
+
+type VolumeAppService struct {
+	source string
+}
+
+func (v *VolumeAppService) Prepare(ctx context.Context, docker *client.Client, contextStr string) error {
+	asVolume, err := docker.VolumeCreate(context.Background(), volume.VolumesCreateBody{
+		Name: "appservices",
+	})
+	if err != nil {
+		return err
+	}
+	v.source = asVolume.Name
+	return nil
+}
+
+func (v *VolumeAppService) Mount() mount.Mount {
+	return mount.Mount{
+		Type:   mount.TypeVolume,
```

Comment:
The changes in this PR seem to break running Complement on my macOS machine: `failed to deployBaseImage: Error response from daemon: invalid mount config for type "": mount type unknown`

 - macOS 11.1
 - Docker desktop 4.3.2 (72729)
 - Docker Engine 20.10.11

```sh
$ COMPLEMENT_ALWAYS_PRINT_SERVER_LOGS=1 COMPLEMENT_DIR=../complement ./scripts-dev/complement.sh TestImportHistoricalMessages
[+] Building 130.6s (18/18) FINISHED
 => [internal] load build definition from Dockerfile                                                                                                                                                                   0.1s
 => => transferring dockerfile: 3.43kB                                                                                                                                                                                 0.0s
 => [internal] load .dockerignore                                                                                                                                                                                      0.2s
 => => transferring context: 181B                                                                                                                                                                                      0.0s
 => [internal] load metadata for docker.io/library/python:3.8-slim                                                                                                                                                     0.6s
 => CACHED [builder 1/9] FROM docker.io/library/python:3.8-slim@sha256:d20122663d629b8b0848e2bb78d929c01aabab37c920990b37bb32bc47328818                                                                                0.0s
 => [internal] load build context                                                                                                                                                                                      0.3s
 => => transferring context: 7.03MB                                                                                                                                                                                    0.2s
 => [builder 2/9] RUN    --mount=type=cache,target=/var/cache/apt,sharing=locked    --mount=type=cache,target=/var/lib/apt,sharing=locked  apt-get update && apt-get install -y     build-essential     libffi-dev    63.6s
 => [builder 3/9] COPY scripts /synapse/scripts/                                                                                                                                                                       0.1s
 => [builder 4/9] COPY MANIFEST.in README.rst setup.py synctl /synapse/                                                                                                                                                0.0s
 => [builder 5/9] COPY synapse/__init__.py /synapse/synapse/__init__.py                                                                                                                                                0.0s
 => [builder 6/9] COPY synapse/python_dependencies.py /synapse/synapse/python_dependencies.py                                                                                                                          0.0s
 => [builder 7/9] RUN --mount=type=cache,target=/root/.cache/pip   pip install --prefix="/install" --no-warn-script-location     /synapse[all]                                                                        56.1s
 => [builder 8/9] COPY synapse /synapse/synapse/                                                                                                                                                                       0.2s
 => [builder 9/9] RUN pip install --prefix="/install" --no-deps --no-warn-script-location /synapse                                                                                                                     4.7s
 => CACHED [stage-1 2/5] RUN    --mount=type=cache,target=/var/cache/apt,sharing=locked    --mount=type=cache,target=/var/lib/apt,sharing=locked   apt-get update && apt-get install -y     curl     gosu     libjpeg  0.0s
 => [stage-1 3/5] COPY --from=builder /install /usr/local                                                                                                                                                              1.3s
 => [stage-1 4/5] COPY ./docker/start.py /start.py                                                                                                                                                                     0.0s
 => [stage-1 5/5] COPY ./docker/conf /conf                                                                                                                                                                             0.1s
 => exporting to image                                                                                                                                                                                                 1.6s
 => => exporting layers                                                                                                                                                                                                1.6s
 => => writing image sha256:23f170fca54b55c6613de47dac9e5c622d5dee4aeb6bf9a1b4a7034a4fdfa1ec                                                                                                                           0.0s
 => => naming to docker.io/matrixdotorg/synapse                                                                                                                                                                        0.0s

Use 'docker scan' to run Snyk tests against images to find vulnerabilities and learn how to fix them
[+] Building 2.6s (11/11) FINISHED
 => [internal] load build definition from Synapse.Dockerfile                                                                                                                                                           0.1s
 => => transferring dockerfile: 891B                                                                                                                                                                                   0.0s
 => [internal] load .dockerignore                                                                                                                                                                                      0.0s
 => => transferring context: 2B                                                                                                                                                                                        0.0s
 => [internal] load metadata for docker.io/matrixdotorg/synapse:latest                                                                                                                                                 0.0s
 => [internal] load build context                                                                                                                                                                                      0.2s
 => => transferring context: 10.48kB                                                                                                                                                                                   0.0s
 => [1/6] FROM docker.io/matrixdotorg/synapse:latest                                                                                                                                                                   0.3s
 => [2/6] COPY synapse/* /conf/                                                                                                                                                                                        0.1s
 => [3/6] COPY keys/* /ca/                                                                                                                                                                                             0.1s
 => [4/6] RUN openssl genrsa -out /conf/server.tls.key 2048                                                                                                                                                            0.5s
 => [5/6] RUN generate_signing_key.py -o /conf/server.signing.key                                                                                                                                                      1.3s
 => [6/6] WORKDIR /data                                                                                                                                                                                                0.1s
 => exporting to image                                                                                                                                                                                                 0.1s
 => => exporting layers                                                                                                                                                                                                0.1s
 => => writing image sha256:8977dd20d4660131c2b8be59dd6cd2c87aaaa858c49262067a0c61fc0b6b388f                                                                                                                           0.0s
 => => naming to docker.io/library/complement-synapse                                                                                                                                                                  0.0s

Use 'docker scan' to run Snyk tests against images to find vulnerabilities and learn how to fix them
Images built; running complement
2022/01/12 14:50:45 config: &{BaseImageURI:complement-synapse BaseImageArgs:[] DebugLoggingEnabled:false AlwaysPrintServerLogs:true BestEffort:false SpawnHSTimeout:30s KeepBlueprints:[] PackageNamespace:pkg}
=== RUN   TestImportHistoricalMessages
2022/01/12 14:50:45 pkg.hs_with_application_service.hs1 : failed to deployBaseImage: Error response from daemon: invalid mount config for type "": mount type unknown
2022/01/12 14:50:45 pkg.hs_with_application_service.hs2 : failed to deployBaseImage: Error response from daemon: invalid mount config for type "": mount type unknown
    msc2716_test.go:63: Deploy: Failed to construct blueprint: Error response from daemon: invalid mount config for type "": mount type unknown
--- FAIL: TestImportHistoricalMessages (0.10s)
FAIL
FAIL	github.com/matrix-org/complement/tests	0.522s
2022/01/12 14:50:45 config: &{BaseImageURI:complement-synapse BaseImageArgs:[] DebugLoggingEnabled:false AlwaysPrintServerLogs:true BestEffort:false SpawnHSTimeout:30s KeepBlueprints:[] PackageNamespace:csapi}
testing: warning: no tests to run
PASS
ok  	github.com/matrix-org/complement/tests/csapi	0.270s [no tests to run]
FAIL
```

Reply:
Looks like the problem is around the `/ca` mount which doesn't have a type defined.

```
mounts="[
  {
    Type:
    Source:/Users/eric/Documents/github/element/complement/tests/ca
    Target:/ca
    ReadOnly:false
    BindOptions:<nil>
    VolumeOptions:<nil>
    TmpfsOptions:<nil>
  }
  {
    Type:volume
    Source:appservices
    Target:/appservices
    ReadOnly:false
    BindOptions:<nil>
    VolumeOptions:<nil>
    TmpfsOptions:<nil>
  }
]"
```

Reply:
See https://github.com/matrix-org/complement/pull/283/files#r783477732 for pointing out the problem

Reply:
Fix at https://github.com/matrix-org/complement/pull/285